contracts/CreditVoucher.sol

```diff
@@ -0,0 +1,255 @@
+     * @param _validityDuration The validity duration to redeem the voucher
+     * @param _deadline The deadline timestamp by which the call must be mined for the approve to work
+     * @param _signature The signature provided by token owner
+     */
+    function mint(
+        address _to,
+        uint256 _creditValue,
+        uint256 _validityDuration,
+        uint256 _deadline,
+        bytes calldata _signature
+    ) public {
+        if (_msgSender() != _to) {
+            revert InvalidMsgSender();
+        }
+
+        // Automatically burn expired vouchers to free up USDC reserves for new minting.
+        uint256 newCursor = batchBurn(burnCursor, MAX_BURN_AMOUNT);
+
+        // Skip updating burn cursor if it hasnt changed.
+        if (burnCursor < newCursor) {
+            burnCursor = newCursor;
+        }
+
+        if (totalCredits + _creditValue > USDC.balanceOf(address(this))) {
+            revert InsufficientUSDCReserve();
+        }
+
+        bytes32 structHash = keccak256(abi.encode(MINT_TYPEHASH, _to, _creditValue, sigNonces[_to]++, _deadline));
+        bytes32 digest = _hashTypedDataV4(structHash);
+        _validateRecoveredAddress(digest, operator, _deadline, _signature);
+
+        uint256 tokenId = tokenCounter++;
+        _safeMint(_to, tokenId);
+        vouchers[tokenId] = Voucher(_creditValue, _validityDuration, block.timestamp, 0, false);
+
+        totalCredits += _creditValue;
+
+        emit Minted(_to, tokenId, _creditValue);
+    }
+
+    /**
+     * @notice Redeem a voucher for its pegged token equivalent
+     * @dev    Caller can be ANYONE
+     * @param _tokenId The tokenId to be redeemed
+     */
+    function redeem(uint256 _tokenId) public {
+        address msgSender = _msgSender();
+        if (ownerOf(_tokenId) != msgSender) {
+            revert NotVoucherOwner();
+        }
+
+        Voucher storage voucher = vouchers[_tokenId];
+        uint256 current = block.timestamp;
+
+        if (voucher.redeemed) {
+            revert VoucherRedeemed();
+        }
+
+        if (current > voucher.createdAt + voucher.validityDuration) {
+            revert VoucherExpired();
+        }
+
+        voucher.redeemedAt = current;
+        voucher.redeemed = true;
+        totalCredits -= voucher.creditValue;
+        _burn(_tokenId);
+
+        IERC20(USDC).transfer(msgSender, voucher.creditValue);
+
+        emit Redeemed(msgSender, _tokenId);
+    }
+
+    /**
+     * @notice Burn an expired voucher
+     * @dev    Caller can be ANYONE
+     * @param _tokenId The tokenId to be burned
+     */
+    function burn(uint256 _tokenId) external {
+        Voucher storage voucher = vouchers[_tokenId];
+
+        if (voucher.redeemed) {
+            revert VoucherRedeemed();
+        }
+
+        if (block.timestamp <= voucher.createdAt + voucher.validityDuration) {
+            revert VoucherUnexpiredYet();
+        }
+
+        totalCredits -= voucher.creditValue;
+        _burn(_tokenId);
```

Comment:
I have been thinking: what do we gain by calling the ERFC721 "_burn" function? 

We could leave the token existing, but in an "expired" state (which can be inferred from the "createdAt" and "validityDuration"). We don't need to burn the token.

In other words, once a token is minted, it never gets burned and is always in one of three states: valid, redeemed, expired. These three states can be inferred from `createdAt`, `redeemedAt` and `validityDuration`.

Would there be any drawback with this approach?